File: CalculateTemplate.m
function rval = CalculateTemplate(inputs)

wavelength = inputs.wavelength;
n_pixels   = inputs.n_pixels;
pixel_p    = inputs.pixel_p;
ref_dist   = inputs.ref_dist;
samp_dist  = inputs.samp_dist;
ref_amp    = inputs.ref_amp;
ref_pos_x  = inputs.ref_pos_x;
ref_pos_y  = inputs.ref_pos_y;
obj_amp    = inputs.obj_amp;

template_pos_x = 0;
template_pos_y = 0;
template_pos_z = samp_dist;
pixel_x        = pixel_p*([0:n_pixels-1] - floor(n_pixels/2));
pixel_y        = pixel_p*([0:n_pixels-1] - floor(n_pixels/2));
ref_pos_z      = ref_dist; % what distance is laser from the sample

% Compute the reference wave on the sensor:
template_phase = 2*pi*sqrt((template_pos_x - ref_pos_x)^2 + (template_pos_y - ref_pos_y)^2 + (template_pos_z - ref_pos_z)^2)/wavelength;
ref_wave       = single(zeros(n_pixels, n_pixels));
template_wave  = single(zeros(n_pixels, n_pixels));
for ix = 1:n_pixels
    % (we're only vectorizing over the y axis to save memory)
    x = pixel_x(ix);

    ref_dist         = sqrt((x - ref_pos_x).^2 + (pixel_y - ref_pos_y).^2 + ref_pos_z.^2); % distance from referece to each pixel
    ref_n_waves      = ref_dist/wavelength;
    ref_wave(ix, :)  = single(ref_amp*exp(j*2*pi*ref_n_waves));

    template_dist        = sqrt((x - template_pos_x).^2 + (pixel_y - template_pos_y).^2 + template_pos_z.^2); % distance from template-reference-object to each pixel
    template_n_waves     = template_dist/wavelength;        
    template_wave(ix, :) = single(obj_amp*exp(j*(2*pi*template_n_waves + template_phase)));
end

% Construct the return-value:
rval.template_wave = template_wave; clear template_wave;
rval.ref_wave      = ref_wave; clear ref_wave;

File: cpp/calculate_template.cpp
#include "calculate_template.h"

#include <cmath>

Outputs calculateTemplate(const Inputs& inputs) {
  double wavelength = inputs.wavelength;
  int n_pixels = inputs.n_pixels;
  double pixel_p = inputs.pixel_p;
  double ref_dist = inputs.ref_dist;
  double samp_dist = inputs.samp_dist;
  int ref_amp = inputs.ref_amp;
  int ref_pos_x = inputs.ref_pos_x;
  int ref_pos_y = inputs.ref_pos_y;
  double obj_amp = inputs.obj_amp;

  double template_pos_x = 0;
double template_pos_y = 0;
double template_pos_z = samp_dist;
cv::Mat pixel_x(1, n_pixels, CV_64F);
cv::Mat pixel_y(1, n_pixels, CV_64F);
for (int i = 0; i < n_pixels; ++i) {
pixel_x.at<double>(i) = pixel_p * (i - std::floor(n_pixels / 2));
pixel_y.at<double>(i) = pixel_p * (i - std::floor(n_pixels / 2));
}
double ref_pos_z = ref_dist;

double template_phase = 2 * M_PI * std::sqrt(std::pow(template_pos_x - ref_pos_x, 2) + std::pow(template_pos_y - ref_pos_y, 2) + std::pow(template_pos_z - ref_pos_z, 2)) / wavelength;
cv::Mat ref_wave(n_pixels, n_pixels, CV_32FC1);
cv::Mat template_wave(n_pixels, n_pixels, CV_32FC1);
for (int ix = 0; ix < n_pixels; ++ix) {
    double x = pixel_x.at<double>(ix);
    cv::Mat ref_dist(n_pixels, n_pixels, CV_64F);
    for (int iy = 0; iy < n_pixels; ++iy) {
        ref_dist.at<double>(iy, ix) = std::sqrt(std::pow(x - ref_pos_x, 2) + std::pow(pixel_y.at<double>(iy) - ref_pos_y, 2) + std::pow(ref_pos_z, 2));
    }
    cv::Mat ref_n_waves = ref_dist / wavelength;
    ref_wave.col(ix) = ref_amp * cv::exp(1i * 2 * M_PI * ref_n_waves);

    cv::Mat template_dist(n_pixels, n_pixels, CV_64F);
    for (int iy = 0; iy < n_pixels; ++iy) {
        template_dist.at<double>(iy, ix) = std::sqrt(std::pow(x - template_pos_x, 2)


    /*cv::Mat ref_dist(n_pixels, n_pixels, CV_64F);
    for (int iy = 0; iy < n_pixels; ++iy) {
      ref_dist.at<double>(iy, ix) = std::sqrt(std::pow(x - ref_pos_x, 2) + std::pow(pixel_y.at<double>(iy) - ref_pos_y, 2) + std::pow(ref_pos_z, 2));
    }
    cv::Mat ref_n_waves = ref_dist / wavelength;
    ref_wave.col(ix) = ref_amp * cv::exp(1i * 2 * M_PI * ref_n_waves);

    cv::Mat template_dist(n_pixels, n_pixels, CV_64F);
    for (int iy = 0; iy < n_pixels; ++iy) {*/
      template_dist.at<double>(iy, ix) = std::sqrt(std::pow(x - template_pos_x, 2) + std::pow(pixel_y.at<double>(iy) - template_pos_y, 2) + std::pow(template_pos_z, 2));
    }
    cv::Mat template_n_waves = template_dist / wavelength;
    template_wave.col(ix) = obj_amp * cv::exp(1i * (2 * M_PI * template_n_waves + template_phase));
  }

  Outputs outputs;
  outputs.template_wave = template_wave;
  outputs.ref_wave = ref_wave;
  return outputs;
}